numpy/core/src/multiarray/descriptor.c

```diff
@@ -3112,6 +3112,13 @@ static PyMethodDef arraydescr_methods[] = {
     {"newbyteorder",
         (PyCFunction)arraydescr_newbyteorder,
         METH_VARARGS, NULL},
+
+    /* for typing; requires python >= 3.9 */
+    #ifdef Py_GENERICALIASOBJECT_H
+    {"__class_getitem__",
+        (PyCFunction)Py_GenericAlias,
+        METH_CLASS | METH_O, NULL},
```

Comment:
For reference: the [Py_GenericAlias docs](https://docs.python.org/3/c-api/typehints.html).

Reply:
Probably, I am just lacking the typing knowledge, but why setting this to some generic (slightly weird?) error better than just leaving it undefined?

Reply:
My initial line of thought was that it would make `hasattr(x, "__class_getitem__")` checks more reliable, but after some further thinking I think a more explicit error message might be more desirable here. The conditionally generated documentation is not exactly helpful either (_e.g._ CircleCI runs on 3.8 it seems).

Reply:
Ah, the thing I utterly overlooked was that you want this on `np.generic` or at least `np.inexact` and then you have to override it.